pkg/controller/grouping/controller.go

```diff
@@ -35,24 +35,45 @@ const (
 	resyncPeriod time.Duration = 0
 )
 
+// eventsCounter is used to keep track of the number of occurrences of an event type. It uses the
+// low-level atomic memory primitives from the sync/atomic package to provide atomic operations
+// (Increment and Load).
+// There is a known-bug on 32-bit architectures for sync/atomic:
+// On ARM, 386, and 32-bit MIPS, it is the caller's responsibility to arrange for 64-bit alignment
+// of 64-bit words accessed atomically. The first word in a variable or in an allocated struct,
+// array, or slice can be relied upon to be 64-bit aligned.
+// As a result, instances of eventsCounter should be allocated when using them in structs; they
+// should not be embedded directly.
+type eventsCounter struct {
```

Comment:
i wonder if we can make use of mutex for eventsCounter and increment under a lock instead of using atomic?

Reply:
Please refer to Quan's original PR. I think he was trying to avoid using a mutex. Now that the functionality is encapsulated in `eventsCounter`, we can change the implementation of `Increment` and `Load` at any time. I don't really have a preference between using a mutex or atomics there. Even with a mutex, it's better to manipulate pointers to eventsCounter as mutex objects can't be copied.